src/components/cylc/tree/tree.js

```diff
@@ -0,0 +1,497 @@
+ */
+function computeCyclePointsStates (cyclePointNodes) {
+  for (const cyclePointNode of cyclePointNodes) {
+    const childStates = []
+    for (const child of cyclePointNode.children) {
+      childStates.push(child.node.state)
+    }
+    cyclePointNode.node.state = extractGroupState(childStates, false)
+  }
+}
+
+/**
+ * A data structure representing the tree used by Cylc.
+ *
+ * Each node has a similar structure, represented as follows:
+ *
+ * <script language="javascript">
+ * {
+ *   id: "some string",
+ *   node: {},
+ *   children: []
+ * }
+ * </script>
+ *
+ * At its first level, there is a single root node, the Workflow node.
+ * Workflows have cycle points as children.
+ *
+ * A cycle point, in its turn, has family proxies children.
+ *
+ * A family proxy can have either other family proxies as children, or
+ * task proxies.
+ *
+ * Task proxies have jobs as children.
+ *
+ * And jobs, finally, have job details as children.
+ *
+ * <script language="javascript">
+ * // This data structure keeps the reference to the root node:
+ * const rootNode = cylcTreeInstance.root;
+ * // and its children can be equally accessed with:
+ * const workflowChildren = rootNode.children
+ * const cyclePoint = workflowChildren[0]
+ * // and so it goes
+ * const familyProxy = cyclePoint.children[0]
+ * </script>
+ *
+ * This data structure also keeps a lookup `Map` object. This map
+ * contains the ID of each node as key, and the node itself as object
+ * reference.
+ *
+ * It means that you can easily access any element in the tree, without
+ * having to iterate and visit each of its parents.
+ *
+ * <script language="javascript">
+ * const jobID = 'kinow|five|20150129T1200+12|foo|1'
+ * const someJob = cycleTreeInstance.lookup.get(jobID)
+ * </script>
+ *
+ * Finally, this data structure class contains methods to:
+ *
+ * - add a node of any type
+ * - update a node of any type
+ * - remove a node of any type
+ *
+ * When a node is added, it gets added to the hierarchical tree (i.e. it
+ * will be added as child of some other node), and also added to the
+ * lookup map.
+ *
+ * When a node is updated, the values of the given node argument in the
+ * function replace the values in the existing element of the map. i.e.
+ * we will find the object in the lookup map, and use Lodash's `merge`.
+ *
+ * And when a node is removed, it gets removed from its parent's `.children`
+ * array, and the node and each of its children get removed from the
+ * lookup map as well.
+ */
+class CylcTree {
+  /**
+   * Create a tree with an initial root node, representing
+   * a workflow in Cylc.
+   *
+   * @param {null | {
+   *   id: string,
+   *   node: Object,
+   *   children: []
+   * }} [workflow]
+   */
+  constructor (workflow) {
+    /**
+     * @type {Map<string, any>}
+     */
+    this.lookup = new Map()
+    if (!workflow) {
+      this.root = {
+        id: '',
+        node: {},
+        children: []
+      }
```

Comment:
I'm guessing this is for testing purposes?

If no workflow is provided then the root node doesn't get added to the lookup?

Reply:
>I'm guessing this is for testing purposes?

Also because there are Vue components using the `.children` element. This way we don't have to add `v-if`/`v-show`. The `beforeRouteLeave` also clears the tree, resetting the `.children` array, so that Vue automatically removes the DOM/VDOM elements for us.

Though it should be possible to remove this I guess, and just check for null/undefined in tests & views/components if that's better.

>If no workflow is provided then the root node doesn't get added to the lookup?

That's right. Only values populated via deltas are added to the lookup.